### PR TITLE
bugs in write nsqw to sqw

### LINKED ISSUE #297: write_nsqw_to_sqw fails in parallel for single bin file
Parallel job, combining tmp files into single sqw file hangs up when tmp files consist of single bin. 

Way to reproduce:
take `Horace\_test\test_change_crystal\test_change_crystal_1.m` test and enable hpc_config.combine_sqw_using = 'mpi_code'
(row 22) 
Better to take herbert framework under windows to see what is happening in parallel workers.

the parallel job, which assembles tmp files together hangs up on writing, despite, seems all messages are received into the data  cache.

### PR DESCRIPTION
**Pull request by abuts**
fixes #297 
Code has been substantially refactored and number of various unit tests has been added. 

Combine_sqw should now be robust

### CODE REVIEW

File: _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_herbert.m [hunk 1]
```diff
@@ -50,9 +50,9 @@
                 test_name = mfilename('class');
             end
             combine_algorithm = 'mpi_code'; % this is what should be tested
-            if is_jenkins && ispc
-                combine_algorithm = 'mex_code'; % disable mpi combine on Jenkins. It is extreamly slow.
-            end
+            %if is_jenkins && ispc
```
Comment (nickbattam-tessella) [inline _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_herbert.m:53]: should this section just be deleted?

File: horace_core/sqw/file_io/@combine_sqw_pix_job/private/read_inputs_send_to_writer_.m [hunk 2]
```diff
@@ -21,39 +16,48 @@
...
+
+        n_source = obj.labIndex-obj.reader_id_shift_;
+        payload = obj.mess_struct_;
+        payload.n_source = n_source;
+        payload.bin_range= [ibin_start,ibin_end];
         
-        payload = struct('lab',obj.labIndex,'messN',niter,'npix',0,...
-            'bin_range',[ibin_start,ibin_end],'pix_tb',[],...
-            'filled_bin_ind',[]);
+%         payload = struct('lab',obj.labIndex,'messN',niter,'npix',0,...
```
Comment (nickbattam-tessella) [inline horace_core/sqw/file_io/@combine_sqw_pix_job/private/read_inputs_send_to_writer_.m:49]: should this section be deleted?

Comment (abuts) [reply horace_core/sqw/file_io/@combine_sqw_pix_job/private/read_inputs_send_to_writer_.m:49]: this certainly shold

File: horace_core/sqw/file_io/class_helpers/pix_cache.m [hunk 3]
```diff
@@ -106,74 +227,114 @@
...
-            end
+            %             if h_log
+            %                 fprintf(h_log,' will save bins: [%d , %d]; ****** saving pixels: %d, pix left: %d\n',...
+            %                     first_bin_to_proc,last_bin_to_process,size(pix_section,2),npix_left);
+            %                 fprintf(h_log,' cache contains: \n');
+            %                 for j=1:n_mess
+            %                     fprintf(h_log,'file %d; bin-range: [%d, %d] n full bins: %d\n',...
+            %                         j,obj.bin_range_(1,j),obj.bin_range_(2,j),numel(obj.filled_bin_ind_{j}));
+            %                 end
+            %             end
```
Comment (hsautessella) [inline horace_core/sqw/file_io/class_helpers/pix_cache.m:335]: Is this block needed any more?

File: horace_core/sqw/file_io/@combine_sqw_pix_job/private/write_npix_to_pix_blocks_.m [hunk 4]
```diff
@@ -45,13 +43,16 @@ function write_npix_to_pix_blocks_(obj,fout,pix_out_position,pix_comb_info)
...
 
+obj.pix_combine_info_ = pix_comb_info;
 
+obj=obj.init_reader_task();
 % Get number of files
-fid = verify_and_reopen_input_files_(pix_comb_info);
+%fid = verify_and_reopen_input_files_(obj);
 % Always close opened files on the procedure completion
-clob = onCleanup(@()fcloser_(fid));  %
+%clob = onCleanup(@()fcloser_(fid));  %
```
Comment (hsautessella) [inline horace_core/sqw/file_io/@combine_sqw_pix_job/private/write_npix_to_pix_blocks_.m:55]: Some more commented code here

**Review by hsautessella — COMMENTED** [review]
Looks fine to me. A couple more comments that look as though they could be removed, but that's all 👍

File: horace_core/sqw/file_io/class_helpers/pix_cache.m (hunk 3 above)
Comment (abuts) [reply horace_core/sqw/file_io/class_helpers/pix_cache.m:335]: probably better to re-enable it, though hope one will never need to use it

File: _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_herbert.m (hunk 1 above)
Comment (abuts) [reply _test/test_gen_sqw_workflow/test_gen_sqw_accumulate_sqw_herbert.m:53]: theoretically, job should not depend on framework, and I have verified that it does not. The filebased MPI is indeed very slow for pixel combining so it probably  better to enable this block again.

File: _test/test_sqw/split_to_messages_for_testing.m [hunk 5]
```diff
@@ -1,35 +1,104 @@
...
+        bin_edges_expanded = zeros(numel(bin_sequence)+1,1);
 
-    n_bin_filled = numel(filled_bin_ind);
-    pix_tb = cell(1,n_bin_filled);    
-    for j=1:n_bin_filled 
-        pix_pos = file_pix(1,:)== filled_bin_ind(j);
-        pix_tb{j} = file_pix(:,pix_pos);
+        n_bin = 1;
+        for j=1:numel(bin_sequence)
+%             if n_bin > numel(flld_bin_ind)
```
Comment (nickbattam-tessella) [inline _test/test_sqw/split_to_messages_for_testing.m:83]: more commented out code

File: _test/test_sqw/split_to_messages_for_testing.m [hunk 6]
```diff
@@ -1,35 +1,107 @@
...
+% last_bin_completed -- if true, the messages contain full bins. If false,
+% some pixels from the bin are not fitting the buffer and will be placed in
+% messages later.
+
 if ~exist('fileind','var')
     fileind = 1:n_files;
 end
+if nargin<6
+    npix1 = find(test_pix_block(1,:)>=nbin_start,1);
+end
```
Comment (cmarooney-stfc) [inline _test/test_sqw/split_to_messages_for_testing.m]: Looks as if npix1 is arg#7 if present. Guessing it is possible for filebuf to be present as an argument but not npix1, so maybe this test should be <7.

Comment (cmarooney-stfc) [reply _test/test_sqw/split_to_messages_for_testing.m]: sorry, buf_size

File: horace_core/sqw/file_io/@combine_sqw_pix_job/combine_sqw_pix_job.m [hunk 7]
```diff
@@ -10,6 +10,10 @@
         is_finished_  = false;
         finalizer_ =[];
         open_files_id_ = [];
+        % The number of auxiliary worker, not participating in reading data
+        % It can be 1 single writer-combiner/mutliple readers
+        % or 2 (1 writer, 1 combiner/multiple readers)
+        reader_id_shift_ = 1;
```
Comment (cmarooney-stfc) [inline horace_core/sqw/file_io/@combine_sqw_pix_job/combine_sqw_pix_job.m]: Is it possible to do named constants or enumerations for the values 1 and 2 here so the code reader can see immediately what has been chosen?

Comment (abuts) [reply horace_core/sqw/file_io/@combine_sqw_pix_job/combine_sqw_pix_job.m]: It is possible, though not very convenient. You roughly define a class, with C-style enumeration (see MESS_CODES.m) or more complex class connecting strings and numbers (config classes). Its not done there, as I suspect there would be no need in keeping both options. I expect that after proper profiling only one job configuration will be proven more performant -- option 2. I may be wrong and if any each configuration will be best for some situation, we will do this and add the appropriate text option to `hpc_config` class. At the moment it is only internal option to change at debugging

File: _test/test_sqw/split_to_messages_for_testing.m (hunk 6 above)
Comment (abuts) [reply _test/test_sqw/split_to_messages_for_testing.m]: the comment is to a previous version of the code. The new one does not have this part

**Pull request merged**